README.md
[![CircleCI](https://circleci.com/gh/ideal-postcodes/ospoint.svg?style=svg)](https://circleci.com/gh/ideal-postcodes/ospoint) [![Coverage Status](https://coveralls.io/repos/github/ideal-postcodes/ospoint/badge.svg?branch=master)](https://coveralls.io/github/ideal-postcodes/ospoint?branch=master) ![Minified & Gzipped Size](https://img.shields.io/bundlephobia/minzip/ospoint.svg?style=flat) ![Dependencies](https://img.shields.io/david/ideal-postcodes/ospoint.svg?style=flat)

# Ordnance Survey Point Converter

[![Try OsPoint on RunKit](https://badge.runkitcdn.com/ospoint.svg)](https://npm.runkit.com/ospoint)

Converts Ordnance Survey grid points (UK National Grid, Northings & Eastings) into latitude and longitude

For anyone who has ever tried to get useful location data from [Ordnance Survey's Code-Point Open](https://www.ordnancesurvey.co.uk/opendatadownload/products.html) but was stuck with Northings and Eastings, you can use this package to convert those numbers into longitude and latitude.

OSPoint will allow you to convert Northings and Eastings into OSGB36, ETRS89 or WGS84 coordinates. If you're not sure which one you need, you probably just want WGS84.

This package is based on equations provided by the people at the Ordance Survey.

OSPoint will accurately translate Northings and Eastings for the United Kingdom. 

**New** Now supports conversion of Irish National Grid coordinates. Please take care when transforming Irish coordinates - you will need to pass in the name of the [proper mercator projection](http://en.wikipedia.org/wiki/Irish_grid_reference_system) (as demonstrated below).

## Getting Started

```javascript
const OSPoint = require('ospoint');

// Create a new OSPoint instance, with Northings & Eastings
const point = new OSPoint("NORTHINGS", "EASTINGS");

// Retrieve OSGB coordinates
point.toOSGB36();

// Retrieve ETRS89 coordinates
point.toETRS89();

// Retrieve WGS84 coordinates
point.toWGS84();
```

If your Northings and Eastings data is a Irish National Grid coordinate, be sure to pass in `"irish_national_grid"` when converting. OSPoint will then use the appropriate mercator projection. For example:

```javascript
const OSPoint = require('ospoint');

// Create a new OSPoint instance, with Irish Northings & Eastings
const point = new OSPoint("NORTHINGS", "EASTINGS");

// Retrieve ETRS89 coordinates
point.toETRS89("irish_national_grid");

// Retrieve WGS84 coordinates
point.toWGS84("irish_national_grid");
```

## Nota Bene

- Northings and Eastings must be a number (no grid letters)
- To be safe, you should pass in raw Northings and Eastings as strings. They occasionally come with a leading 0s from the OS dataset, which javascript will interpret as a base-8 number
- All transformation methods return an object with longitude and latitude properties in decimal degrees

## Testing

```shell
npm test
```

## Note on ETRS89 and WGS84 Coordinate Systems

ETRS89 is a variation of WGS84 that takes into account the slow North Easterly drifting of the Eurasian tectonic plate. The WGS84 and ETRS89 coordinate systems coincided in 1989 (hence the name) and have drifted apart at a rate of ~2.5cm per year due to tectonic movements.

toWGS84() is currently hard-coded to output the same coordinates as toETRS89(). But in reality WGS and ETRS are [off by ~50-60cm in 2013](http://www.killetsoft.de/t_1009_e.htm).

## Which Coordinate System do I want to translate Northings and Eastings into?

95% of the time, people want WGS84

## License

MIT



CHANGELOG.md
# Changelog

Any changes, including backwards incompatible changes will be listed here
   
## 0.2.1 (11/02/2019)

- Produce smaller build sizes by dropping extraneous package artefacts



tests/unit.js
var	assert = require('chai').assert,
		OSPoint = require('../index.js');
		roughEquals = require("./helpers").equal,

describe("OSPoint", function () {
	it ("should return correct eastings and northings", function () {
		var	northings= 12345678,
				eastings= 12345678;
				point = new OSPoint(northings, eastings);
		assert.equal(point.northings, northings);
		assert.equal(point.eastings, eastings);
	});
});

describe("#toDecimalFromDMS", function () {
	it ("should correctly convert to decimal from DMS", function () {
		assert.equal(0, OSPoint.toDecimalFromDMS(0, 0, 0));
		assert.equal(30, OSPoint.toDecimalFromDMS(30, 0, 0));
		assert.equal(roughEquals(OSPoint.toDecimalFromDMS(15,0,3), 15.00083333333333), true);
	});
});

describe("#toRadians", function () {
	it("should convert degrees to radians", function () {
		assert.equal(roughEquals(OSPoint.toRadians(180), Math.PI), true);
		assert.equal(roughEquals(OSPoint.toRadians(0), 0), true);
		assert.equal(roughEquals(OSPoint.toRadians(90), Math.PI / 2), true);
	});
});

describe("#toDegrees", function () {
	it ("should convert radians to degrees", function () {
		assert.equal(roughEquals(OSPoint.toDegrees(Math.PI), 180), true);
		assert.equal(roughEquals(OSPoint.toDegrees(0), 0), true);
		assert.equal(roughEquals(OSPoint.toDegrees(Math.PI / 2), 90), true);
	})
})

describe("#nu", function () {
	it ("should return the correct value of nu", function () {
		// Airy 1830 Ellipsoid, National Grid example
		var a = 6377563.396, 
				F0 = 0.9996012717,
				e2 = OSPoint.e2(a, 6356256.909), 
				lat = OSPoint.toRadians(OSPoint.toDecimalFromDMS(52, 39, 27.2531)),
				expected = 6388502.3333,
				result = OSPoint.nu(a, F0, e2, lat);
		assert.equal(roughEquals(expected, result), true);
	});
});

describe("#rho", function () {
	it ("should return the correct value of rho", function () {
		// Airy 1830 Ellipsoid, National Grid example
		var a = 6377563.396,
				F0 = 0.9996012717,
				e2 = OSPoint.e2(a, 6356256.909), 
				lat = OSPoint.toRadians(OSPoint.toDecimalFromDMS(52, 39, 27.2531)),
				expected = 6372756.4399,
				result = OSPoint.rho(a, F0, e2, lat);
		assert.equal(roughEquals(expected, result), true);
	});
});

describe("#eta2", function () {
	it ("should return the correct value of eta squared", function () {
		// Airy 1830 Ellipsoid, National Grid example
		var a = 6377563.396,
				F0 = 0.9996012717,
				e2 = OSPoint.e2(a, 6356256.909), 
				lat = OSPoint.toRadians(OSPoint.toDecimalFromDMS(52, 39, 27.2531)),
				rho = OSPoint.rho(a, F0, e2, lat),
				nu = OSPoint.nu(a, F0, e2, lat),
				expected = 0.0024708136169,
				result = OSPoint.eta2(nu, rho);
		assert.equal(roughEquals(expected, result), true);
	});
});

describe("#M", function () {
	it ("should return the correct value of M", function () {
		// Airy 1830 Ellipsoid, National Grid example
		var a = 6377563.396,
				b = 6356256.909,
				F0 = 0.9996012717,
				lat = OSPoint.toRadians(OSPoint.toDecimalFromDMS(52, 39, 27.2531)),
				lat0 = OSPoint.toRadians(OSPoint.toDecimalFromDMS(49, 0, 0)),
				expected = 406688.29596,
				result = OSPoint.M(a, b, F0, lat, lat0);
		assert.equal(roughEquals(expected, result), true);
	});
});

describe("#mercatorToLatLon", function () {
	it ("should return the correct longitude and latitude", function () {
		// Airy 1830 Ellipsoid, National Grid example
		var NORTHINGS = 313177.270,
				EASTINGS = 651409.903,
				expected_lat = OSPoint.toDecimalFromDMS(52, 39, 27.2531),
				expected_lon = OSPoint.toDecimalFromDMS(1, 43, 4.5177),
				result = new OSPoint(NORTHINGS, EASTINGS).mercatorToLatLon();
		assert.equal(roughEquals(OSPoint.toDegrees(result.longitude), expected_lon), true);
		assert.equal(roughEquals(OSPoint.toDegrees(result.latitude), expected_lat), true);
	});
});

describe("#toCartesian", function () {
	it ("should correctly map lon, lat & height to 3D cartesian points", function () {
		var	lon = OSPoint.toRadians(OSPoint.toDecimalFromDMS(1, 43, 4.5177)),
				lat = OSPoint.toRadians(OSPoint.toDecimalFromDMS(52, 39, 27.2531)),
				height = 24.7,
				expected_x = 3874938.849,
				expected_y = 116218.624,
				expected_z = 5047168.208,
				result = OSPoint.toCartesian(lon, lat, height);
		assert.equal(roughEquals(expected_x, result.x), true);
		assert.equal(roughEquals(expected_y, result.y), true);
		assert.equal(roughEquals(expected_z, result.z), true);
	});
});

describe("#toLatLon", function () {
	it ("should correctly map 3d cartiesian points to lon, lat & height given ellipsoid", function () {
		var	expected_lon = OSPoint.toDecimalFromDMS(1, 43, 4.5177),
				expected_lat = OSPoint.toDecimalFromDMS(52, 39, 27.2531),
				expected_height = 24.7,
				x = 3874938.849,
				y = 116218.624,
				z = 5047168.208,
				result = OSPoint.toLatLon(x, y ,z);
		assert.equal(roughEquals(expected_lat, OSPoint.toDegrees(result.latitude)), true);
		assert.equal(roughEquals(expected_lon, OSPoint.toDegrees(result.longitude)), true);
		assert.equal(expected_height - result.height < 0.001, true); // To nearest mm=
	});
});


tests/helpers/index.js
var TOLERANCE = 0.00001;

exports.equal = function (a, b) {
	absA = Math.abs(a);
	absB = Math.abs(b);
	var diff = Math.abs(a - b);

	if (a == b) {
		return true;
	} else if (a * b == 0) {
		return diff < TOLERANCE * TOLERANCE;
	} else {
		return (diff / (absA + absB)) < TOLERANCE;
	}
}


tests/national_grid_unit.js
	var fs = require("fs"),
			path = require("path"),
			OSPoint = require("../index"),
			assert = require("chai").assert,
			roughEquals = require("./helpers").equal,
			testDataPath = path.join(__dirname, "data/national_grid.json"),
			testData;

var testData = JSON.parse(fs.readFileSync(testDataPath)).data;

describe("Transformation", function () {	
	it ("should have the right OSGB36 conversion", function () {
		var testPoint, osgb36, t;
		for (var i = 0; i < testData.length; i += 1) {
			t = testData[i];
			testPoint = new OSPoint(t.northings, t.eastings);
			osgb36 = testPoint.toOSGB36();
			assert.equal(roughEquals(osgb36.latitude, t.osgb36.latitude), true);
			assert.equal(roughEquals(osgb36.longitude, t.osgb36.longitude), true);
		}
	});
	it ("should have the right ETRS89 conversion", function () {
		var testPoint, etrs89, t;
		for (var i = 0; i < testData.length; i += 1) {
			t = testData[i];
			testPoint = new OSPoint(t.northings, t.eastings);
			etrs89 = testPoint.toETRS89();
			assert.equal(roughEquals(etrs89.latitude, t.etrs89.latitude), true);
			assert.equal(roughEquals(etrs89.longitude, t.etrs89.longitude), true);
		}
	});

	it ("should have the right WGS84 conversion", function () {
		var testPoint, wgs84, t;
		for (var i = 0; i < testData.length; i += 1) {
			t = testData[i];
			testPoint = new OSPoint(t.northings, t.eastings);
			wgs84 = testPoint.toWGS84();
			assert.equal(roughEquals(wgs84.latitude, t.etrs89.latitude), true);
			assert.equal(roughEquals(wgs84.longitude, t.etrs89.longitude), true);
		}
	});
});

index.js
/*
*
* OSPoint
*	Based on Equations Provided by Ordnance Survey:
*	A Guide to Coordinate Systems in Great Britain (Dec 2010, D00659 v2.1)
* https://github.com/cblanc/ospoint
*
*/

/* 
 * UTILITY FUNCTIONS 
 * These functions have been factored out and named to assist readability
 * when used in conjunction with the equations provided by Ordnance Survey
 *
 */

"use strict";

function OSPoint (northings, eastings) {
	this.northings = parseInt(northings, 10),
	this.eastings = parseInt(eastings, 10);
}

OSPoint.toRadians = function (degrees) {
	return toRadians(degrees);
}

OSPoint.toDegrees = function (radians) {
	return toDegrees(radians);
}

OSPoint.nu = function (a, F0, e2, lat) {
	var x = a * F0,
			y = 1 - (e2 * sin2(lat));

	return (x * Math.pow(y, -0.5));
}

OSPoint.rho = function (a, F0, e2, lat) {
	var x = a * F0,
			y = 1 - e2,
			z = 1 - (e2 * sin2(lat));
	return x * y * Math.pow(z, -1.5);
}

OSPoint.e2 = function (a, b) {
	return (a*a - b*b) / (a*a);
}

OSPoint.M = function (a, b, F0, lat, lat0) {
	var n = (a - b) / (a + b),
			n2 = n * n,
			n3 = n * n * n;

	var w = (1 + n + (5/4)*n2 + (5/4)*n3) * (lat - lat0),
			x = (3*n + 3*n2 + (21/8)*n3) * sin(lat - lat0) * cos(lat + lat0),
			y = ((15/8)*n2 + (15/8)*n3) * sin(2*(lat -lat0)) * cos(2*(lat + lat0)),
			z = (35/24)*n3 * sin(3*(lat - lat0)) * cos(3*(lat+lat0));
	return b * F0 * (w - x + y - z);
}

OSPoint.eta2 = function (nu, rho) {
	return (nu / rho) - 1;
}

OSPoint.toDecimalFromDMS = function (degrees, minutes, seconds) {
	return toDecimalFromDMS(degrees, minutes, seconds);
}

var sin = function (radians) {
	return Math.sin(radians);
}

var sin2 = function (radians) {
	return Math.pow(Math.sin(radians), 2);
}

var cos = function (radians) {
	return Math.cos(radians);
}

var tan = function (radians) {
	return Math.tan(radians);
}

var tan2 = function (radians) {
	return Math.pow(Math.tan(radians), 2);
}

var tan4 = function (radians) {
	return Math.pow(Math.tan(radians), 4);
}

var tan6 = function (radians) {
	return Math.pow(Math.tan(radians), 6);
}

var arctan = function (radians) {
	return Math.atan(radians);
}

var sec = function (radians) {
	return Math.pow(Math.cos(radians), -1);
}

var toRadians = function (degrees) {
	return (degrees / 180) * Math.PI;
}

var toDegrees = function (radians) {
	return (radians / Math.PI) * 180;
}

var toDecimalFromDMS = function (degrees, minutes, seconds) {
	return degrees + (minutes / 60) + (seconds / 3600);
}

var toRadiansFromDMS = function (degrees, minutes, seconds) {
	return toRadians(toDecimalFromDMS(degrees, minutes, seconds));
}

/*
 *
 * Real Work Starts Here
 *
 */

var ELLIPSOID = {
	airy1830: {
		a: 6377563.396, // Semi-major axis
		b: 6356256.909  // Semi-minor axis
	},
	grs80: {
		a: 6378137.000,
		b: 6356752.3141
	},
	airy1830_modified: {
		a: 6377340.189,
		b: 6356034.447
	}
};

var MERCATOR_PROJECTIONS = {
	national_grid: {
		scaleFactor: 0.9996012717,
		trueOrigin: {
			longitude: toRadiansFromDMS(-2, 0, 0),
			latitude: toRadiansFromDMS(49, 0, 0),
			eastings: 400000,
			northings: -100000
		},
		ellipsoid: 'airy1830'
	},
	irish_national_grid: {
		scaleFactor: 1.000035,
		trueOrigin: {
			longitude: toRadiansFromDMS(-8, 0, 0),
			latitude: toRadiansFromDMS(53, 30, 0),
			eastings: 200000,
			northings: 250000
		},
		ellipsoid: 'airy1830_modified'
	}
};

OSPoint.prototype.toOSGB36 = function (projection) {
	// Use default projection
	var result = mercatorToLatLon(this.northings, this.eastings, projection);
	return {
		longitude: OSPoint.toDegrees(result.longitude),
		latitude: OSPoint.toDegrees(result.latitude)
	};
}

OSPoint.prototype.toETRS89 = function (projection) {
	var osgb36 = mercatorToLatLon(this.northings, this.eastings, projection),
			height = 0,
			osgbTransform = OSPoint.toCartesian(osgb36.longitude, osgb36.latitude, height),
			point = OSPoint.helmertTransformation(osgbTransform),
			result = OSPoint.toLatLon(point.x, point.y, point.z, 'grs80');
	return {
		longitude: OSPoint.toDegrees(result.longitude),
		latitude: OSPoint.toDegrees(result.latitude)
	};
}

OSPoint.prototype.toWGS84 = function (projection) {
	// Set to ETRS89
	return this.toETRS89(projection);
}

OSPoint.prototype.mercatorToLatLon = function () {
	return mercatorToLatLon(this.northings, this.eastings);
}

OSPoint.toCartesian = function (lon, lat, H, ellipsoid) {
	var ellipsoid = ELLIPSOID[ellipsoid] || ELLIPSOID['airy1830'],
			a = ellipsoid.a,
			b = ellipsoid.b,
			e2 = OSPoint.e2(a, b),
			v = a * Math.pow(1 - e2 * sin2(lat), -0.5),
			x = (v + H) * cos(lat) * cos(lon),
			y = (v + H) * cos(lat) * sin(lon),
			z = ((1-e2)*v + H) * sin(lat);

	return {
		x: x,
		y: y,
		z: z
	};
}

OSPoint.toLatLon = function (x, y, z, ellipsoid) {
	var ellipsoid = ELLIPSOID[ellipsoid] || ELLIPSOID['airy1830'],
			a = ellipsoid.a,
			b = ellipsoid.b,
			e2 = OSPoint.e2(a, b),
			lon = arctan(y / x),
			p = Math.pow(x*x + y*y, 0.5),
			initialLat = arctan(z / (p * (1-e2) )),
			estimateV = function (lat) {
				return a * Math.pow(1 - e2 * sin2(lat), -0.5);
			},
			v = estimateV(initialLat),
			estimateLat = function (v, prevLat) {
				return arctan( (z + e2*v*sin(prevLat)) / p);
			},
			lat = estimateLat(v, initialLat),
			TOLERANCE = 0.0000000001;

	while (Math.abs(lat - initialLat) > TOLERANCE) {
		v = estimateV(lat);
		initialLat = lat;
		lat = estimateLat(v, initialLat);
	}

	var height = (p / cos(lat)) - v;

	return {
		latitude: lat,
		longitude: lon,
		height: height
	};
}

/*
 * The Helmert Transformation, a reference
 *
 * [Xb]   [tx]   [ 1+s -rz ry  ]   [Xa]
 * [Yb] = [ty] + [ rz  1+s -rx ] * [Ya]
 * [Zb]   [tz]   [ -ry rx  1+s ]   [Za]
 *
 */

var HELMERT_TRANSFORMATIONS = {
	etrs89: {
		to_osgb36: {
			tx: -446.448,		// metres
			ty: 125.157,		// metres
			tz: -542.060,		// metres
			rx: -0.1502,		// seconds
			ry: -0.2470,		// seconds
			rz: -0.8421,		// seconds
			s: 20.4894			// ppm
		}
	},
	osgb36: {
		to_etrs89: {
			tx: 446.448,
			ty: -125.157,
			tz: 542.060,
			rx: 0.1502,
			ry: 0.2470,
			rz: 0.8421,
			s: -20.4894
		}
	}
};

OSPoint.helmertTransformation = function (point, transformation) {
	var t = transformation || HELMERT_TRANSFORMATIONS.osgb36.to_etrs89,
			xa = point.x,
			ya = point.y,
			za = point.z,
			tx = t.tx,
			ty = t.ty,
			tz = t.tz,
			rx = OSPoint.toRadians(t.rx / 3600),
			ry = OSPoint.toRadians(t.ry / 3600),
			rz = OSPoint.toRadians(t.rz / 3600),
			s  = t.s * 0.000001,
			xb = tx + (1+s)*xa + (-rx*ya) + (ry * za),
			yb = ty + rz*xa + (1+s)*ya + (-rx)*za,
			zb = tz + (-ry)*xa + rx*ya + (1+s)*za;

	return {
		x: xb,
		y: yb,
		z: zb
	};
}

var mercatorToLatLon = function (northings, eastings, projection) {
	var projection = MERCATOR_PROJECTIONS[projection] || MERCATOR_PROJECTIONS['national_grid'],
			ellipsoid = ELLIPSOID[projection.ellipsoid],
			E = eastings,
			N = northings,
			a = ellipsoid.a,
			b = ellipsoid.b,
			e2 = OSPoint.e2(a, b),
			F0 = projection.scaleFactor,
			N0 = projection.trueOrigin.northings,
			E0 = projection.trueOrigin.eastings,
			lat0 = projection.trueOrigin.latitude,
			lon0 = projection.trueOrigin.longitude,
			latDashed = ((N - N0) / (a * F0)) + lat0, // Equation C6
			M = OSPoint.M(a, b, F0, latDashed, lat0);

	while (N - N0 - M >= 0.0001) {
		latDashed = ((N - N0 - M) / (a * F0)) + latDashed; // Equation C7
		M = OSPoint.M(a, b, F0, latDashed, lat0);	
	}
	
	var rho = OSPoint.rho(a, F0, e2, latDashed),
			nu = OSPoint.nu(a, F0, e2, latDashed),
			eta2 = OSPoint.eta2(nu, rho),

			VII = function (latDashed, nu, rho) {
				return tan(latDashed) / (2 * rho * nu);
			}(latDashed, nu, rho),

			VIII = function (latDashed, nu, rho, eta2) {
				var x = tan(latDashed) / (24 * rho * Math.pow(nu, 3)),
						y = 5 + 3*tan2(latDashed) + eta2 - (9*tan2(latDashed)*eta2);
				return x * y;
			}(latDashed, nu, rho, eta2),

			IX = function (latDashed, nu, rho) {
				var x = tan(latDashed) / (720 * rho * Math.pow(nu, 5)),
						y = 61 + 90*tan2(latDashed) + 45*tan4(latDashed);
				return x * y;
			}(latDashed, nu, rho),

			X = function (latDashed, nu) {
				return sec(latDashed) / nu;
			}(latDashed, nu),

			XI = function (latDashed, nu ,rho) {
				var x = sec(latDashed) / (6 * Math.pow(nu, 3)),
						y = (nu / rho) + (2 * tan2(latDashed));
				return x * y;
			}(latDashed, nu ,rho),

			XII = function (latDashed, nu) {
				var x = sec(latDashed) / (120 * Math.pow(nu, 5)),
						y = 5 + 28*tan2(latDashed) + 24*tan4(latDashed);
				return x * y;
			}(latDashed, nu),

			XIII = function (latDashed, nu) {
				var x = sec(latDashed) / (5040 * Math.pow(nu, 7)),
						y = 61 + 662*tan2(latDashed) + 1320*tan4(latDashed) + 720*tan6(latDashed);
				return x * y;
			}(latDashed, nu),
			E = E - E0,
			E2 = E  * E,
			E3 = E2 * E,
			E4 = E3 * E,
			E5 = E4 * E,
			E6 = E5 * E,
			E7 = E6 * E,
			lat = latDashed - VII*E2 + VIII*E4 - IX*E6,
			lon = lon0 + X*E - XI*E3 + XII*E5 - XIII*E7;
	
	return {
		latitude: lat,
		longitude: lon
	};
}

module.exports = OSPoint;
